sgd/src/main/java/RunSGD.java

```diff
@@ -0,0 +1,244 @@
+        final DataQuantaBuilder<?, double[]> weightsBuilder = javaPlanBuilder.loadCollection(broadcast).withName("init weights");
+
+        final DataQuantaBuilder<?, SparseVector> transformBuilder = javaPlanBuilder
+                .readTextFile(fileName).withName("source")
+                .map(new Transform()).withName("transform");
+
+        Collection<double[]> results  =
+                weightsBuilder.doWhile(new LoopCondition(accuracy, max_iterations), w -> {
+
+            DataQuantaBuilder<?, double[]> newWeightsDataset = transformBuilder
+                    .sample(sampleSize)
+                    .map(new ComputeLogisticGradient()).withBroadcast(w, "weights").withName("compute")
+                    .reduce(new Reduce()).withName("reduce")
+                    .map(new WeightsUpdate()).withBroadcast(w, "weights").withName("update");
+
+            DataQuantaBuilder<?, Tuple2<Double, Double>> convergenceDataset = newWeightsDataset.map(new Converge()).withBroadcast(w, "weights");
+
+            return new Tuple<>(newWeightsDataset.filter(w1 -> true), convergenceDataset);
+        }).collect();
+
+        System.out.println("Output weights:" + Arrays.toString(RheemCollections.getSingle(results)));
+
+    }
+}
+
+class Transform implements FunctionDescriptor.SerializableFunction<String, SparseVector> {
+
+    @Override
+    public SparseVector apply(String line) {
+        String[] pointStr = line.split(" ");
+        int[] indices = new int[pointStr.length - 1];
+        double[] values = new double[pointStr.length - 1];
+        double label = Double.parseDouble(pointStr[0]);
+        for (int i = 1; i < pointStr.length; i++) {
+            if (pointStr[i].equals("")) {
+                continue;
+            }
+            String kv[] = pointStr[i].split(":", 2);
+            indices[i - 1] = Integer.parseInt(kv[0]);
+            values[i - 1] = Double.parseDouble(kv[1]);
+        }
+        return new SparseVector(label, indices, values);
+    }
+}
+
+class ComputeLogisticGradient implements FunctionDescriptor.ExtendedSerializableFunction<SparseVector, double[]> {
+
+    double[] weights;
+
+    @Override
+    public double[] apply(SparseVector point) {
+        double[] gradient = new double[weights.length + 1];
+        double dot = 0;
+        for (int i = 0; i < point.size(); i++)
+            dot += weights[point.getIndices()[i] - 1] * point.getValues()[i];
+
+        for (int i = 0; i < point.size(); i++)
+            gradient[point.getIndices()[i]] = ((1 / (1 + Math.exp(-1 * dot))) - point.getLabel()) * point.getValues()[i];
+
+        gradient[0] = 1; //counter required in the update
+
+        return gradient;
+    }
+
+    @Override
+    public void open(ExecutionContext executionContext) {
+        this.weights = (double[]) executionContext.getBroadcast("weights").iterator().next();
+    }
+}
+
+class Reduce implements FunctionDescriptor.SerializableBinaryOperator<double[]> {
```

Comment:
`Reduce` and `Converge` are a bit generic names, so it's not immediately clear what they are doing.